subt/cloudsim2osgar.py

```diff
@@ -58,21 +60,45 @@ def publish(self, channel, data):
 
 
 class main:
-    def __init__(self):
+    def __init__(self, robot_name):
         # get cloudsim ready
-        robot_name = sys.argv[1]
-        imu_name = '/'+robot_name+'/imu/data'
         wait_for_master()
-        rospy.init_node('imu2osgar', log_level=rospy.DEBUG)
-        rospy.loginfo("waiting for {}".format(imu_name))
-        rospy.wait_for_message(imu_name, Imu)
-        rospy.sleep(2)
-        self.imu_count = 0
-
-        # start
+        rospy.init_node('cloudsim2osgar', log_level=rospy.DEBUG)
         self.bus = Bus()
-        self.bus.register('rot', 'acc', 'orientation')
-        rospy.Subscriber(imu_name, Imu, self.imu)
+
+        # common topics
+        topics = [
+            ('/' + robot_name + '/imu/data', Imu, self.imu, ('rot', 'acc', 'orientation')),
+        ]
+
+        # configuration specific topics
+        robot_description = rospy.get_param("/{}/robot_description".format(robot_name))
+        if "robotika_x2_sensor_config_1" in robot_description:
+            rospy.loginfo("robotika x2")
+        elif "ssci_x2_sensor_config_1" in robot_description:
+            rospy.loginfo("ssci x2")
+        elif "ssci_x4_sensor_config_2" in robot_description:
+            rospy.loginfo("ssci drone")
+            topics.append(('/' + robot_name + '/top_scan', LaserScan, self.top_scan, ('top_scan',)))
+            topics.append(('/' + robot_name + '/bottom_scan', LaserScan, self.bottom_scan, ('bottom_scan',)))
+        elif "TeamBase" in robot_description:
+            rospy.loginfo("teambase")
+        elif "robotika_freyja_sensor_config_1" in robot_description:
+            rospy.loginfo("freya 1")
+        else:
+            rospy.logerror("unknown configuration")
```

Comment:
Well, I am not sure about this - so for every new robot (even Freyja 2) we will have to modify, `run_solution.bash`, add launch files, maybe JSON and also this `cloudsim2osgar`. Is it necessary? I originally planed to add exception only for SSCI X4.

Reply:
Either way when we add rgbd cameras they are different on each of the robots. So we would have to add more xml files to remap topics to the same names. I'd rather keep the readability by keeping everything contained here in one file, to be able to read it top to bottom and get the whole picture.

The error at the end is meant for anyone adding a new configuration, to check this file and by adding one more `elif` to confirm that nothing needs changing (which will never be the case I am afraid).